wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/LanguageServerConfig.java

```diff
@@ -0,0 +1,189 @@
+ */
+public interface LanguageServerConfig {
+
+  /**
+   * Get a language server regular expression provider.
+   *
+   * @return regular expression provider
+   */
+  RegexProvider getRegexpProvider();
+
+  /**
+   * Get a language server communication provider. If there is no need for a custom communication
+   * provider, there is a default implementation {@link EmptyCommunicationProvider}.
+   *
+   * @return communication provider
+   * @see EmptyCommunicationProvider
+   */
+  CommunicationProvider getCommunicationProvider();
+
+  /**
+   * Get a language server instance provider. If there is no need for a custom instance provider,
+   * there is a default implementation {@link DefaultInstanceProvider}
+   *
+   * @return instance provider
+   * @see DefaultInstanceProvider
+   */
+  InstanceProvider getInstanceProvider();
+
+  /**
+   * Get an installer status provider.
+   *
+   * @return <code>null</code>if a language server is not expected to be launched locally, instance
+   *     of an installer status provider otherwise
+   */
+  default InstallerStatusProvider getInstallerStatusProvider() {
+    return null;
+  }
+
+  /**
+   * Instance provider is the definition of the way a language server instance is constructed. It is
+   * not restricted to provide an instance in any possible way, however the library that is uses
+   * <code>lsp4j</code> imposes its restrictions. The implementer must use {@link LanguageClient}
+   * and {@link LanguageServer} interfaces correspondingly.
+   *
+   * @see DefaultInstanceProvider
+   */
+  interface InstanceProvider {
+    /**
+     * Get a new language server instance for specified communication means.
+     *
+     * @param client language server client
+     * @param in language server input stream
+     * @param out language server output stream
+     * @return language server instance
+     */
+    LanguageServer get(LanguageClient client, InputStream in, OutputStream out);
+  }
+
+  /**
+   * Provides some way we can communicate with the running language server. In general the
+   * implementer has no obligation as of how to launch/initialize local or remote language server
+   * process/server, it can be done separately and it is not required that this interface covers the
+   * procedure, however it is also not disallowed.
+   *
+   * <p>{@link SocketCommunicationProvider} and {@link ProcessCommunicationProvider} implementations
+   * of the interface, besides providing communication means establishes a socket connection and
+   * launches a language server process correspondingly.
+   *
+   * @see EmptyCommunicationProvider
+   * @see SocketCommunicationProvider
+   * @see ProcessCommunicationProvider
+   */
+  interface CommunicationProvider {
+    /**
+     * Get input stream
+     *
+     * @return input stream
+     * @throws LanguageServerException if any issue happened trying to get an input stream
+     */
+    InputStream getInputStream() throws LanguageServerException;
```

Comment:
In general, it is implementation specific, so I prefer to keep it the way it is right now. Your suggestion is limited to cases where they create a new instance, not reuse the old one. Perhaps there is a need to mention that in a javadoc.

Reply:
Or may be not. You see, your suggestion limits the scope to only creating of an entity, while in general it may be not created but obtained already existing one. Example:
In `org.eclipse.che.api.languageserver.EmptyCommunicationProvider#getInputStream` the stream is created, while in `org.eclipse.che.api.languageserver.SocketCommunicationProvider#getInputStream` the stream is not created, instead we get already existing one.

Reply:
You need to define whether that method is called once (as in a factory) or multiple times. If it is called multiple times, the implementer may need to manage the stream instance. If it's called a single time, it's a factory and thus misnamed.